### ISSUE TITLE
DI: Using generics in Angular 2.0

### ISSUE THREAD

**Issue by matanlurey** [angular/angular#1691]
In Angular 1.0 Dart, it is possible to use a TypeLiteral to bind and inject generics:

```dart
class TypeLiteral<T> {
  const TypeLiteral();

  Type get type => t;
}
```

An example of use:
```dart
@Component(
  module: MyComponent.module
)
class MyComponent {
  static void module(DirectiveBinder directiveBinder) {
    directiveBinder.bind(
        const TypeLiteral<List<String>>().type,
        toValue: <String> const ['Hello', 'World']);
  }

  MyComponent(List<String> stringList) {
    assert(stringList == const ['Hello', 'World']);
  }
}
```

In Angular 2, [based on this snippet](https://github.com/angular/angular/blob/master/modules/angular2/src/di/binding.js#L467) Tim linked me to, it looks like that is no longer possible. Can we add this is a FR for NG 2.0/DI? If not, is there an alternative that will also play well with Angular 1.0 Dart's DI system?

@kegluneq, @yjbanov

**Comment by vicb** [comment]
@matanlurey IIRC (the DI repo might have more info) `TypeLiteral` was a workaround of some bugs in the mirrors.

While looking at this, it would be great to check if the following is now possible:
```dart
    directiveBinder.bind(
        List<String>, // no more need for TypeLiteral<> ?
        toValue: <String> const ['Hello', 'World']);
```

**Comment by matanlurey** [comment]
I am 100% sure that is not possible, you would still need TypeLiteral to get accurate typing.

As you said, in static mode, it may or may not be necessary.

**Comment by matanlurey** [comment]
After talking offline with @yjbanov, DI would need to support the following:

```dart
@Component(
  injectables: const [
    const Binding(const TypeLiteral<List<String>>(), toFactory: ...)
  ]
)
class MyComponent {}
```

When Angular DI sees a `TypeLiteral`, they would invoke the `.type` getter to get the `Type`:

```dart
/// Workaround in Dart for getting the true "type" of a generic.
/// E.g. `const TypeLiteral<List<String>>().type`
class TypeLiteral<T> {
  const TypeLiteral();

  Type get type => T;
}
```

@kegluneq for reference.

**Comment by yjbanov** [comment]
@kevmoo, is this something that could be made easier on Dart side?

**Comment by naomiblack** [comment]
Internal tracking: see https://b2/20945733

**Comment by kevmoo** [comment]
This is a pending language request: https://code.google.com/p/dart/issues/detail?id=11923

**Comment by kevmoo** [comment]
We may have a duplicate here: https://code.google.com/p/dart/issues/detail?id=23221

**Comment by yjbanov** [comment]
fixed by #1779
